### ISSUE TITLE
Lenovo: Can't update firmware on Lenovo T480s

### ISSUE THREAD

**Issue by tpopela** [fwupd/fwupd#1101]
Can't update the firmware on the T480s. There are updates available, but when I run `fwupdmgr update` the update is scheduled, but when restarting the OS is boot normally (no updater is invoked).

```shell
0 $ fwupdmgr get-updates
20L8S2N80D System Firmware has firmware updates:
GUID:                    ebfe8df8-dee7-4692-a721-cbcf5095c5cf
ID:                      com.lenovo.ThinkPadN22ET.firmware
Update Version:          0.1.29
Update Name:             ThinkPad T480s
Update Summary:          Lenovo ThinkPad T480s System Firmware
Update Remote ID:        lvfs
Update Checksum:         SHA1(646d495a92291b4832225183ace8b88dd8e9ae88)
Update Location:         https://fwupd.org/downloads/7d2e5ce0403a04d9cdb73e5607f94d6b51e4f1e7-Lenovo-ThinkPad-T480s-SystemFirmware-1.29.cab
Update Description:      Lenovo ThinkPad T480s System Firmware Version 1.29

                         This update also adds the following features:

ID:                      com.lenovo.ThinkPadN22ET.firmware
Update Version:          0.1.27
Update Name:             ThinkPad T480s
Update Summary:          Lenovo ThinkPad T480s System Firmware
Update Remote ID:        lvfs
Update Checksum:         SHA1(a074dd91e0988c3460edf0de75d843d808491bea)
Update Location:         https://fwupd.org/downloads/d3a8889c8ec5985311addbb6fe022495b5120ca8-Lenovo-ThinkPad-T480s-SystemFirmware-1.27.cab
Update Description:      Lenovo ThinkPad T480s System Firmware Version 1.27

                         This update also adds the following features:

No upgrades for UEFI Device Firmware, current is 184.55.3510: 184.55.3510=same
```

```shell
tpopela  ~  >  13:04
0 $ fwupdmgr update
Downloading 0.1.29 for 20L8S2N80D System Firmware...
Decompressing…           [***************************************]
Authenticating…          [***************************************]
Updating 20L8S2N80D System Firmware…                             ]
Scheduling…              [***************************************]
No upgrades for UEFI Device Firmware, current is 184.55.3510: 184.55.3510=same

An update requires a reboot to complete. Restart now? [Y|n]: n
```

```shell
$ fwupdmgr --version
client version: 1.2.5
compile-time dependency versions
        gusb:   0.3.0
        efivar: 37
daemon version: 1.2.5
```

```shell
$ fwupdmgr get-devices
Thinkpad T480s Thunderbolt Controller
  DeviceId:             010926bac4e8994820c0064541587cff5064685a
  Guid:                 3b125a20-5ed2-5c0f-b22d-ab0e0b687bbc
  Summary:              Unmatched performance for high-speed I/O
  Plugin:               thunderbolt
  Flags:                internal|updatable|registered
  Vendor:               Lenovo
  VendorId:             TBT:0x0109
  Version:              14.00
  Icon:                 computer
  Created:              2019-03-18

20L8S2N80D System Firmware
  DeviceId:             e7cc234e82d08b6a14e246beb6b5f8142091e570
  Guid:                 ebfe8df8-dee7-4692-a721-cbcf5095c5cf
  Plugin:               uefi
  Flags:                internal|updatable|require-ac|supported|registered|needs-reboot
  Checksum:             SHA1(3392441eeed2db57dc6bd61ab5eba5eef67bfd83)
  Checksum:             SHA256(fa9b9485b6bc13c4c599d52d8dd6af4af110557f363eee5e822604157b0a10b3)
  Version:              0.1.26
  VersionLowest:        0.0.1
  Icon:                 computer
  Created:              2019-03-18
  UpdateState:          success

UEFI Device Firmware
  DeviceId:             118b3d8cfafa8781890c4d3e23a4deb9fcb18e5a
  Guid:                 39a39e45-38ce-4b05-809b-60351bf88ec4
  Plugin:               uefi
  Flags:                internal|updatable|require-ac|supported|registered|needs-reboot
  Version:              184.55.3510
  VersionLowest:        0.0.1
  Icon:                 audio-card
  Created:              2019-03-18

UEFI Device Firmware
  DeviceId:             323bda90831520867f8c02544efb9711dfca9fb6
  Guid:                 4d254d6e-cd67-477b-97d5-bc3048af45c4
  Plugin:               uefi
  Flags:                internal|updatable|require-ac|registered|needs-reboot
  Version:              0.1.8
  VersionLowest:        0.1.8
  Icon:                 audio-card
  Created:              2019-03-18

SAMSUNG MZVLW256HEHP-000L7
  DeviceId:             c82980cd3fe66f7d0a8bb83b353cbc271177efd4
  Guid:                 5b3df2da-f745-5fd0-81de-5dafd7f0bf8c
  Guid:                 8fd4ca73-d0ae-52e8-8977-461435c6f4cf
  Guid:                 257d6faa-82f3-53e2-afa4-f6b9adc9595d
  Summary:              NVM Express Solid State Drive
  Plugin:               nvme
  Flags:                internal|updatable|require-ac|registered|needs-reboot
  Vendor:               Samsung Electronics Co Ltd
  VendorId:             NVME:0x144D
  Version:              5L7QCXB7
  Icon:                 drive-harddisk
  Created:              2019-03-18
```

```shell
$ efibootmgr -v
BootNext: 0001
BootCurrent: 0000
Timeout: 0 seconds
BootOrder: 0000,0017,0018,0019,001A,001B,001C,001D,001E,0001
Boot0000* Fedora        HD(1,GPT,1456979b-5a85-436c-a15e-3d567e8cceea,0x800,0x200000)/File(\EFI\fedora\shimx64.efi)
Boot0001* Linux-Firmware-Updater        HD(1,GPT,1456979b-5a85-436c-a15e-3d567e8cceea,0x800,0x200000)/File(\EFI\fedora\shimx64.efi)\.f.w.u.p.d.x.6.4...e.f.i...
Boot0010  Setup FvFile(721c8b66-426c-4e86-8e99-3457c46ab0b9)
Boot0011  Boot Menu     FvFile(126a762d-5758-4fca-8531-201a7f57f850)
Boot0012  Diagnostic Splash Screen      FvFile(a7d8d9a6-6ab0-4aeb-ad9d-163e59a7a380)
Boot0013  Lenovo Diagnostics    FvFile(3f7e615b-0d45-4f80-88dc-26b234958560)
Boot0014  Startup Interrupt Menu        FvFile(f46ee6f4-4785-43a3-923d-7f786c3c8479)
Boot0015  Rescue and Recovery   FvFile(665d3f60-ad3e-4cad-8e26-db46eee9f1b5)
Boot0016  MEBx Hot Key  FvFile(ac6fd56a-3d41-4efd-a1b9-870293811a28)
Boot0017* USB CD        VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,86701296aa5a7848b66cd49dd3ba6a55)
Boot0018* USB FDD       VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,6ff015a28830b543a8b8641009461e49)
Boot0019* NVMe0 VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,001c199932d94c4eae9aa0b6e98eb8a400)
Boot001A* ATA HDD0      VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,91af625956449f41a7b91f4f892ab0f600)
Boot001B* USB HDD       VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,33e821aaaf33bc4789bd419f88c50803)
Boot001C* PCI LAN       VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,78a84aaf2b2afc4ea79cf5cc8f3d3803)
Boot001D  Other CD      VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,aea2090adfde214e8b3a5e471856a35406)
Boot001E  Other HDD     VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,91af625956449f41a7b91f4f892ab0f606)
Boot001F* USBR BOOT CDROM       PciRoot(0x0)/Pci(0x14,0x0)/USB(11,1)
Boot0020* USBR BOOT Floppy      PciRoot(0x0)/Pci(0x14,0x0)/USB(11,0)
Boot0021* ATA HDD       VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,91af625956449f41a7b91f4f892ab0f6)
Boot0022* ATAPI CD      VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,aea2090adfde214e8b3a5e471856a354)
Boot0023* PCI LAN       VenMsg(bc7838d2-0f82-4d60-8316-c068ee79d25b,78a84aaf2b2afc4ea79cf5cc8f3d3803)
```

```shell
$ efivar -l | grep fw
0abba7dc-e516-4167-bbf5-4d9d1c739416-fwupd-ebfe8df8-dee7-4692-a721-cbcf5095c5cf-0
```

```shell
$ tree /boot
/boot
├── config-4.20.13-200.fc29.x86_64
├── config-4.20.14-200.fc29.x86_64
├── config-4.20.15-200.fc29.x86_64
├── efi
│   ├── EFI
│   │   ├── BOOT
│   │   │   ├── BOOTIA32.EFI
│   │   │   ├── BOOTX64.EFI
│   │   │   ├── fbia32.efi
│   │   │   └── fbx64.efi
│   │   └── fedora
│   │       ├── BOOTIA32.CSV
│   │       ├── BOOTX64.CSV
│   │       ├── fonts
│   │       │   └── unicode.pf2
│   │       ├── fw
│   │       │   ├── fwupd-3b8c8162-188c-46a4-aec9-be43f1d65697.cap
│   │       │   └── fwupd-ebfe8df8-dee7-4692-a721-cbcf5095c5cf.cap
│   │       ├── fwupdx64.efi
│   │       ├── gcdia32.efi
│   │       ├── gcdx64.efi
│   │       ├── grub.cfg
│   │       ├── grubenv
│   │       ├── grubia32.efi
│   │       ├── grubx64.efi
│   │       ├── mmia32.efi
│   │       ├── mmx64.efi
│   │       ├── shim.efi
│   │       ├── shimia32.efi
│   │       ├── shimia32-fedora.efi
│   │       ├── shimx64.efi
│   │       └── shimx64-fedora.efi
│   ├── mach_kernel
│   └── System
│       └── Library
│           └── CoreServices
│               └── SystemVersion.plist
├── elf-memtest86+-5.01
├── extlinux
│   ├── cat.c32
│   ├── chain.c32
│   ├── cmd.c32
│   ├── cmenu.c32
│   ├── config.c32
│   ├── cptime.c32
│   ├── cpu.c32
│   ├── cpuid.c32
│   ├── cpuidtest.c32
│   ├── debug.c32
│   ├── dhcp.c32
│   ├── dir.c32
│   ├── disk.c32
│   ├── dmi.c32
│   ├── dmitest.c32
│   ├── elf.c32
│   ├── ethersel.c32
│   ├── gfxboot.c32
│   ├── gpxecmd.c32
│   ├── hdt.c32
│   ├── hexdump.c32
│   ├── host.c32
│   ├── ifcpu64.c32
│   ├── ifcpu.c32
│   ├── ifmemdsk.c32
│   ├── ifplop.c32
│   ├── kbdmap.c32
│   ├── kontron_wdt.c32
│   ├── ldlinux.c32
│   ├── lfs.c32
│   ├── libcom32.c32
│   ├── libgpl.c32
│   ├── liblua.c32
│   ├── libmenu.c32
│   ├── libutil.c32
│   ├── linux.c32
│   ├── ls.c32
│   ├── lua.c32
│   ├── mboot.c32
│   ├── memdisk
│   ├── meminfo.c32
│   ├── menu.c32
│   ├── pci.c32
│   ├── pcitest.c32
│   ├── pmload.c32
│   ├── poweroff.c32
│   ├── prdhcp.c32
│   ├── pwd.c32
│   ├── pxechn.c32
│   ├── reboot.c32
│   ├── rosh.c32
│   ├── sanboot.c32
│   ├── sdi.c32
│   ├── sysdump.c32
│   ├── syslinux.c32
│   ├── vesa.c32
│   ├── vesainfo.c32
│   ├── vesamenu.c32
│   ├── vpdtest.c32
│   ├── whichsys.c32
│   └── zzjson.c32
├── grub2
│   ├── grubenv -> ../efi/EFI/fedora/grubenv
│   └── themes
│       └── system
│           ├── background.png
│           └── fireworks.png
├── initramfs-0-rescue-7ccf529d5a9148798a1cb710d47e954a.img
├── initramfs-4.20.13-200.fc29.x86_64.img
├── initramfs-4.20.14-200.fc29.x86_64.img
├── initramfs-4.20.15-200.fc29.x86_64.img
├── loader
│   └── entries
├── lost+found
├── memtest86+-5.01
├── System.map-4.20.13-200.fc29.x86_64
├── System.map-4.20.14-200.fc29.x86_64
├── System.map-4.20.15-200.fc29.x86_64
├── vmlinuz-0-rescue-7ccf529d5a9148798a1cb710d47e954a
├── vmlinuz-4.20.13-200.fc29.x86_64
├── vmlinuz-4.20.14-200.fc29.x86_64
└── vmlinuz-4.20.15-200.fc29.x86_64

16 directories, 105 files
```

Please answer the following questions:

- Operating system and version: Fedora 29
- How did you install fwupd (ex: `from source`, `pacman`, `apt-get`, etc): dnf
- Have you tried rebooting? Yes
- Are you using an NVMe disk? Yes
- Is secure boot enabled (only for the UEFI plugin)? Yes

**Comment by hughsie** [comment]
Can you try doing `/usr/libexec/fwupd/fwupdate --set-debug`, then `fwupdmgr update` then reboot, then on failure `/usr/libexec/fwupd/fwupdate --log` -- thanks!

**Comment by tpopela** [comment]
The --log part didn't worked for me Richard..

1 $ sudo /usr/libexec/fwupd/fwupdate --log
failed: Error opening file /sys/firmware/efi/efivars/FWUPDATE_DEBUG_LOG-0abba7dc-e516-4167-bbf5-4d9d1c739416: No such file or directory

**Comment by hughsie** [comment]
If you press F12 to get the boot menu, can you manually run `fwupdx64.efi`?

**Comment by tpopela** [comment]
No, as there is no entry like that. There's only Fedora, then the SSD drive and then LAN. Also just to be sure I see that in the BIOS overview screen, there is:

UEFI Secure boot: off

but in the secure boot settings it's set to enabled. I don't know whether it can cause something so I'm mentioning it.

**Comment by floriankisser** [comment]
It's working on my T480s with fedora. Secure boot is not an issue and if you have fedora too (assuming yes from your profile pic) I would recommend to turn it on.

Is Security -> "UEFI BIOS Update Option" -> "Windows UEFI Firmware Update" enabled in your BIOS?

**Comment by tpopela** [comment]
We debugged it with Richard on IRC and we didn't found a solution so far.

**Comment by mscherer** [comment]
I do have the same symptom on the T470s of my employer. And to answer @floriankisser , yup, that's "enabled" on my laptop.

**Comment by Peetz0r** [comment]
I was having the same issue on my ThinkPad Yoga 370, but I worked around it and found and worked around some other issues in the process. These are my steps:

* Attempt `fwupdmgr update` a few times, nothing happened
* Manually run `fwupdx64.efi` from a grub shell using chainloading
* Flashed the new firmware
* Fedora uefi entry disapeared. Fallback option didn't work, it bootlooped.
* Got a spare laptop, made a Fedore live usb, booted to a grub shell from there, could boot via there
* Tried chainloading the fallback uefi image (`EFI/BOOT/BOOTX64.EFI`), got an error message (might post this later)
* Tried manually re-creating the fedora uefi entry. Everytime it got removed after a reboot.
* Did a bios settings reset
* Manually recreating uefi entries now worked, also the default Fedora entry got put back by itself somehow.
* Tried re-installing the firmware update (downloaded the .cab, used `--allow-reinstall`), and everything just worked out of the box this time.

I'm running Fedora 29 on a ThinkPad Yoga 370, started the day with firmware 1.32 and now 1.33, fwupdmgr is version 1.2.6.

**Comment by Dapid** [comment]
I have a T480s too, and the upgrade worked without problem. I am running Fedora 30 and fwupmgr 1.2.9.

**Comment by DavidCWGA** [comment]
I have a T480s and I had the same problem. I reset all the BIOS settings to factory defaults, and it started to work.

I haven't pinned down which setting breaks it yet, but the defaults seem to be OK.

**Comment by tpopela** [comment]
Thank you @DavidCWGA that helped me as well and I'm finally up2date (but I didn't took the photo of my previous settings).

**Comment by Franr** [comment]
Was struggling with the same problem, until I figured out that locking the ability of changing the boot order through the BIOS was preventing the firmware updates to boot and start (which make sense!).

**Comment by andodet** [comment]
I am facing the same issues with a t480s (on Linux Mint 20). Running the following commands don't apply any updates at reboot. 
```
sudo fwupdmgr refresh
sudo fwupdmgr get-updates
sudo fwupdmgr update
sudo reboot
```
I am wondering what's the best option to correctly apply updates:

- Wait for Ubuntu to update to a newer snapshot (as mentioned [here][1]).
- Apply [updates][2] manually with `fwupdmgr install *.cab`
- Change BIOS settings (not quite clear what's the one affecting boot order

Thanks everyone for surfacing this.

[1]: https://github.com/fwupd/fwupd/issues/1794#issuecomment-599721939
[2]: https://fwupd.org/lvfs/devices/com.lenovo.ThinkPadN22RN.firmware

**Comment by romunov** [comment]
In case you have a dual boot (windows first) and you reset BIOS you may lose grub from UEFI. At this point you won't be able to boot into Fedora. Good luck!

I solved this problem by 

0) raging why the hell do I have to bother with same problems as in 1998
1) going to the supermarket and buying a USB key
2) making it bootable with supergrub2
3) changed boot sequence in BIOS to boot from the USB key
4) detected boot and selected shim.efi
5) followed instructions from [here](https://superuser.com/questions/596317/how-would-i-reinstall-the-grub-efi-bootloader-on-fedora-linux/596318#596318)

tl;dr

```
sudo efibootmgr -c -L Fedora -d /dev/sda -p 100 -l \\EFI\\fedora\\shim.efi
```
and -d and -p parameters can be found in `lsblk`. Search for `/boot/efi` partition, i.e. `-d /dev/nvme0n1 -p 5`.
